docker/docker_client.go

```diff
@@ -208,9 +209,23 @@ func newDockerClientWithDetails(sys *types.SystemContext, registry, username, pa
 		return nil, err
 	}
 
-	if sys != nil && sys.DockerInsecureSkipTLSVerify {
-		tr.TLSClientConfig.InsecureSkipVerify = true
+	// Check if TLS verification shall be skipped (default=false) which can
+	// either be specified in the sysregistriesv2 configuration or via the
+	// SystemContext, whereas the SystemContext is prioritized.
+	skipVerify := false
+	if sys != nil && sys.DockerInsecureSkipTLSVerify != types.OptionalBoolUndefined {
+		// Only use the SystemContext if the actual value is defined.
+		skipVerify = sys.DockerInsecureSkipTLSVerify == types.OptionalBoolTrue
+	} else {
+		reg, err := sysregistriesv2.FindRegistry(sys, remoteName)
```

Comment:
`remoteName` is `reference.Path(ref)`, i.e. something like `library/busybox`; or it may be `""` with other callers.

I assume this should end up with `ref.Name()` when called by `newDockerClientFromRef`.  But there is also `CheckAuth` and `SearchRegistry`, which only have the hostname.

And, worse, for `docker.io` we have three different strings (`dockerHostname`, `dockerV1Hostname`, and `dockerRegistry`) in various variables, at various points in various code paths.

I’m not sure how to handle this entirely cleanly. Yet another parameter to `newDockerClientWithDetails`?

Reply:
> And, worse, for docker.io we have three different strings (dockerHostname, dockerV1Hostname, and dockerRegistry) in various variables, at various points in various code paths.

I am not sure how we should deal with that. There is a varying degree of craziness in how Docker deals with. `index.docker.io` is mapped to `docker.io` while `registry-1.docker.io` is treated differently (with respect to the RepoTag of the image) and all effectively point to the same image.

```terminal
$ docker pull busybox
$ docker pull library/busybox
$ docker pull docker.io/busybox
$ docker pull docker.io/library/busybox
$ docker pull index.docker.io/busybox
$ docker pull index.docker.io/library/busybox
$ docker pull registry-1.docker.io/library/busybox 

$ docker images
REPOSITORY                             TAG      IMAGE ID [...]
busybox                                latest   f9b6f7f7b9d3 [...]
registry-1.docker.io/library/busybox   latest   f9b6f7f7b9d3 [...]
```

I hope this is something we can figure out at a later point and put it on our todo list?

> I’m not sure how to handle this entirely cleanly. Yet another parameter to newDockerClientWithDetails?

I fear it depends on how we want to deal with `Registry.Prefix` and how we "expect" a registry to be configured. Let's consider the following example (simplified config):
```
url="example.com"
prefix="example.com/foo"
insecure=true
```
If we use `FindRegistry("example.com/bar")` what is the expected outcome? Currently, it wouldn't return any `Registry` as the prefix doesn't match although a user might expect example.com is implicitly set to be insecure? I think there are various (valid?) interpretations of it and I don't think that's good.

I think that TLS verification should always be host/registry/url-based but that doesn't work with the current design of sysregsv2, prefixes in particular, so I want to propose a change that should solve the upper problem.

### Make Prefixes a sub-table
An exemplary config:
```toml
[[registry]]
url = "registry.com"
insecure = true
unqualified-search = true
[[registry.mirror]]
url = "mirror-1.registry.com"
[[registry.prefix]]
path="/bar"
prefix="registry.com/foo"
[[registry.prefix]]
path="/paste"
prefix="registry.com/copy"
trim=true
[[registry.prefix]]
prefix="another-registry.com:5000"
```

The idea is to detach the prefix (and namespace translation) from the registry to a sub-table as already done with mirrors. With the upper example:

- `pull registry.com/foo/image:latest` -> `registry.com/bar/foo/image:latest`
- `pull registry.com/copy/image:latest` -> `registry.com/paste/image:latest` (the prefix is trimmed  before being prepended)
- `pull another-registry.com:5000/image:latest` -> `registry.com/image:latest` (no path specified -> `path="/"`)

@mtrmac This would mean yet another change to `sysregsv2` but, to be honest, I haven't thought about this particular problem. We can add `registry.prefix.trim` at a later point but it should demonstrate that we are more flexible with prefixes being a sub-table. This way, the sysregsv2 configuration is way less ambiguous than before, and we have clear separation between registry-specific config switches and the namespace translations induced by prefixes.

Reply:
### Hostname Mapping

>> And, worse, for docker.io we have three different strings (`dockerHostname, `dockerV1Hostname`, and `dockerRegistry`) in various variables, at various points in various code paths.
>
> I am not sure how we should deal with that. There is a varying degree of craziness in how Docker deals with that. `index.docker.io` is mapped to `docker.io` while `registry-1.docker.io` is treated differently (with respect to the RepoTag of the image) and all effectively point to the same image.

It’s not _that_ bad actually.  For _images_ we support only the syntax `docker.io/*` (via `dockerClientFromRef` and `c/image/docker/reference.ParseNormalizedNamed`, which may make the `docker.io[/library]` prefixes implicit). I don’t think we need to normalize image names to support any of the other values; _users_ always say `docker.io` (or nothing when it is implied), the other `docker.io` hostnames are only set inside this one file.

The underlying `registry-1.docker.io` host name is mapped inside `newDockerClientWithDetails`, so handling that case is just a matter of being careful to use the right variable when calling `FindRegistry`, or something like that.

The V1 `index.docker.io` hostname is used only in the `SearchRegistry` case, which, like `CheckAuth`, works with hostnames only and does not have the `namespace…/repo` subnamespaces.

So, just making sure that `newDockerClientWithDetails` actually looks up the correct registry namespace for all three callers will force the function interface to somehow represent either a hostname-only or a hostname/path input to `FindRegistry`, and at that point the V1/`index.`/`docker.io` distinctions will be reasonably easy to handle.

---

### “Insecure” Flag Duplication.

> If we use FindRegistry("example.com/bar") what is the expected outcome? Currently, it wouldn't return any Registry as the prefix doesn't match although a user might expect example.com is implicitly set to be insecure? I think there are various (valid?) interpretations of it and I don't think that's good.

I’m not _too_ worried about this; for the most part, either a registry hostname is “real” (an underlying server that we want to reach), in which case there will be no configuration or at most a single entry for the whole server; or the hostname is intended to not be used directly, and is mirrored locally (e.g. a local mirror of a tiny subset of `docker.io`, with `Prefix` entries `docker.io/$repo` and `URL` or `Mirrors` pointing to the local mirror), in which case there may be many entries but possibly pointing at different mirrors, and there’s no reason to assume that all the mirrors will use TLS in the same way.

I.e. I’d expect that there is either no entry for `example.com[/anything]`, a single entry for all of `example.com`, or several entries for `example.com/foo` which are likely to match; a request for `example.com/bar` which does not match, but with an existing `example.com/foo` configuration seems unlikely to me.

Still, sure, not repeating the same boolean many times for a single hostname would be simpler, especially when trying to _turn off_ the `insecure` flag.

---

### Prefix Trimming

> We can add `registry.prefix.trim` at a later point but it should demonstrate that we are more flexible with prefixes being a sub-table. 

I don’t think that’s actually necessary; we can always trim the prefix from the match, and the user can undo that by repeating the trimmed part in the URL (i.e. configure `path=/bar/foo` in the `prefix=registry.com/foo` example).

---

### Changing the structure

> I think that TLS verification should always be host/registry/url-based but that doesn't work with the current design of sysregsv2, prefixes in particular, so I want to propose a change that should solve the upper problem.

\*shrug\* We can do so, but in that case let’s get as close to a clean solution as we can.

```
[[registry]]
url = "registry.com"
insecure = true
unqualified-search = true
[[registry.mirror]]
url = "mirror-1.registry.com"
[[registry.prefix]]
path="/bar"
prefix="registry.com/foo"
[[registry.prefix]]
path="/paste"
prefix="registry.com/copy"
trim=true
[[registry.prefix]]
prefix="another-registry.com:5000"
```

That eliminates the duplication of `insecure` and `unqualified-search`, sure; OTOH I’m not sure whether the `blocked` flag is intended to be per-hostname or per-prefix (@rhatdan ?).

WRT usability, it seems a bit confusing the way the mapping for `another-registry.com:5000` is hidden nested somewhere inside the `registry.com` entries; the way I think about this, the “prefix” is the primary lookup key and should be prominent.

Syntactically, the `prefix=[hostname]/path` form is ambiguous because namespace components within a registry can contain dots as well.

Also, I’m not entirely sure that there is a meaningful distinction between the `prefix`→`URL` mapping and between having mirrors; am I overlooking an essential difference?


Can we maybe talk about the configuration as conceptual data structures first, and then figure out the best way to represent it in TOML?

Is this the data we are trying to represent?
```go
type CompleteConfig struct {
  // May be empty. NOTE: Order matters
  UnqualifiedSearchOrder []Hostname
  // Order does not matter
  HostnameConfig map[Hostname]struct {
    // Defaults to false if not specified or there is no match for a Hostname
    Insecure bool
  }
  // Order does not matter as such, but longest match wins. If we match on a longer prefix, the entries for shorter prefixes, if any, are ignored.
  Prefixes map[HostnameRepoPrefix]struct {
    // Defaults to false if not specified or there is no HostnameRepoPrefix match.
    Blocked bool 
    // Defaults to the key of this structure (i.e. identity mapping) if not specified or there is no HostnameRepoPrefix match. Must not be empty.
    PhysicalLocations []HostnameRepoPrefix 
  }
}
```

Reply:
> Syntactically, the `prefix=[hostname]/path` form is ambiguous because namespace components within a registry can contain dots as well.

I’m sorry, actually, if the `/path` variant has a mandatory leading `/`, there is no ambiguity.

Reply:
> I’m not too worried about this; for the most part, either a registry hostname is “real” (an underlying server that we want to reach), in which case there will be no configuration or at most a single entry for the whole server; or the hostname is intended to not be used directly, and is mirrored locally (e.g. a local mirror of a tiny subset of docker.io, with Prefix entries docker.io/$repo and URL or Mirrors pointing to the local mirror), in which case there may be many entries but possibly pointing at different mirrors, and there’s no reason to assume that all the mirrors will use TLS in the same way.
>
>I.e. I’d expect that there is either no entry for example.com[/anything], a single entry for all of example.com, or several entries for example.com/foo which are likely to match; a request for example.com/bar which does not match, but with an existing example.com/foo configuration seems unlikely to me.
>
>Still, sure, not repeating the same boolean many times for a single hostname would be simpler, especially when trying to turn off the insecure flag.

Alright, you convinced me, let's push things forward as they are now. I guess I was overthinking the issue.

Reply:
> I’m not sure how to handle this entirely cleanly. Yet another parameter to newDockerClientWithDetails?

I reduced the list of parameters now as those can (or should?) be set by the callers. However, it's a bit of a deja vu as we run into the same problem we hit in `sysregistriesv2` when parsing strings that can be a fully-qualified image reference, or a registry or a registry plus path. For now, I tried to solve that be having two parameters (i..e, registry and reference) where the reference is used in `FindRegistry(..)`.

Reply:
> >I’m not too worried about this; for the most part, either a registry hostname is “real” (an underlying server that we want to reach), in which case there will be no configuration or at most a single entry for the whole server; or the hostname is intended to not be used directly, and is mirrored locally (e.g. a local mirror of a tiny subset of docker.io, with Prefix entries docker.io/$repo and URL or Mirrors pointing to the local mirror), in which case there may be many entries but possibly pointing at different mirrors, and there’s no reason to assume that all the mirrors will use TLS in the same way.
> 
> Alright, you convinced me, let's push things forward as they are now. I guess I was overthinking the issue.

(FWIW the “not too worried about this” comment was about the narrow case of having a configuration for prefix example.com/foo and using example.com/bar with no configuration.  I do agree about the general ugliness of having possibly several “insecure TLS” flags for the single hostname — it’s not a _blocker_ issue for me, but it is a bit worrying.)

Reply:
>  I reduced the list of parameters now as those can (or should?) be set by the callers.

OK… it’s now easier for a caller to forget to set one up, but on the plus side the values have names, so it’s definitely not worse.